### PR TITLE
MM-15670 Mark as Unread

### PR DESCRIPTION
**Pull request by hmhealey**
This is the PR to merge the mark-as-unread feature branch into master. It's made up of the following PRs:
- #11794 Add the initial API endpoints to mark a channel as unread
- #11932 Fix marking a channel as unread to do it from the correct post since it was previously from the post after the one you selected
- #11966 Refactors some of the notification and mention counting logic and uses that to set the mention badge when marking a channel as unread
- #12074 Adds a websocket event when marking a channel as unread to alert other web/desktop sessions
- #12499 Adds a push notification when marking a channel as unread to alert mobile sessions

#### Ticket Link
https://mattermost.atlassian.net/browse/MM-15670

#### Related PRs
Mobile PR: https://github.com/mattermost/mattermost-mobile/pull/3554
Redux PR: https://github.com/mattermost/mattermost-redux/pull/986
Web PR: https://github.com/mattermost/mattermost-webapp/pull/4213

### CODE REVIEW

**Comment by hmhealey** [comment]
@jgilliam17 To make this testable as a whole feature, I'm going to update the test server that has all of these changes together. I'll post a link to it once that's done

**Comment by hmhealey** [comment]
Also, I'm assigning this to myself so that I can merge it without squashing the feature branch

**Comment by hmhealey** [comment]
@jgilliam17 The test server at https://mark-as-unread.test.spinmint.com is now up to date

File: app/channel.go [hunk 1]
```diff
@@ -1782,6 +1782,40 @@ func (a *App) UpdateChannelLastViewedAt(channelIds []string, userId string) *mod
...
+	if err != nil {
+		return nil, err
+	}
+
+	unreadMentions, err := a.countMentionsFromPost(user, post)
+	if err != nil {
+		return nil, err
+	}
+
+	channel, updateErr := a.Srv.Store.Channel().UpdateLastViewedAtPost(post, userID, unreadMentions)
```
Comment (jespino) [inline app/channel.go]: This looks like is returning a channel, but the returned object is not a channel, can we use another name to make it more clear?

Comment (hmhealey) [reply app/channel.go]: Yep, will do

**Comment by esethna** [comment]
Removing QA label, Jelena on the feature branch spinmint tested

**Comment by deanwhillier** [comment]
@jespino, looks like your requested change was addressed, can you please re-review asap?

File: app/notification_push.go [hunk 2]
```diff
@@ -238,6 +205,32 @@ func (a *App) ClearPushNotification(currentSessionId, userId, channelId string)
 	}
 }
 
+func (a *App) UpdateMobileAppBadgeSync(userId string) *model.AppError {
+	msg := &model.PushNotification{
+		Type:             model.PUSH_TYPE_UPDATE_BADGE,
+		Version:          model.PUSH_MESSAGE_V2,
+		Sound:            "none",
```
[the remark this answers is not in the document]
Comment (hmhealey) [reply app/notification_push.go]: Good catch. I don't think it used a constant originally, and I'm pretty sure I added one just for that

**Pull request merged**